core/src/main/java/com/scalar/db/transaction/consensuscommit/CommitMutationComposer.java

```diff
@@ -99,4 +103,40 @@ private Delete composeDelete(Operation base, TransactionResult result) {
                 new ConditionalExpression(
                     STATE, toStateValue(TransactionState.DELETED), Operator.EQ)));
   }
+
+  private Key getPartitionKey(Operation base, @Nullable TransactionResult result) {
+    if (base instanceof Mutation) {
+      // for usual commit
+      return base.getPartitionKey();
+    } else {
+      assert base instanceof Selection;
+      if (result != null) {
+        // for rollforward in lazy recovery
+        return result.getPartitionKey().get();
+      } else {
+        // for deleting non-existing record that was prepared with DELETED for Serializable with
+        // Extra-write
+        assert base instanceof Get;
+        return base.getPartitionKey();
+      }
+    }
+  }
+
+  private Optional<Key> getClusteringKey(Operation base, @Nullable TransactionResult result) {
+    if (base instanceof Mutation) {
+      // for usual commit
+      return base.getClusteringKey();
+    } else {
+      assert base instanceof Selection;
+      if (result != null) {
+        // for rollforward in lazy recovery
+        return result.getClusteringKey();
+      } else {
+        // for deleting non-existing record that was prepared with DELETED for Serializable with
+        // Extra-write
+        assert base instanceof Get;
+        return base.getClusteringKey();
+      }
+    }
+  }
```

Comment:
This is the logic to get partition key and clustering key for `CommitMutationComposer`.